06_gpu_and_ml/text-to-video/mochi.py

```diff
@@ -28,338 +27,147 @@
-)
-
-
-@app.function(
-    volumes={MODEL_CACHE: model},
-    timeout=2 * HOURS,
-    image=download_image,
-)
-def download_model(
-    model_revision: str = "8e9673c5349979457e515fddd38911df6b4ca07f",
-):
-    from huggingface_hub import snapshot_download
-    from transformers import T5EncoderModel, T5Tokenizer
+MINUTES = 60
+HOURS = 60 * MINUTES
 
-    model.reload()
-    print("🍡 downloading Mochi model")
+# ## Setting up our Mochi class
+# We'll use the `modal.cls` decorator to define a [Modal Class](https://modal.com/docs/guide/lifecycle-functions)
+# which we use to control the lifecycle of our cloud container.
+# 
+# We configure it to use our image, the distributed volume, and a single H100 GPU.
 
-    snapshot_download(
-        repo_id="genmo/mochi-1-preview",
-        local_dir=MODEL_CACHE / "mochi-1-preview",
-        revision=model_revision,
+with image.imports():
+    import torch
+    from diffusers import MochiPipeline
+    from diffusers.utils import export_to_video
+    
+@app.cls(
+    image=image,
+    volumes={
+        OUTPUTS_PATH: outputs,  # videos will be saved to a distributed volume
+    },
+    gpu=modal.gpu.H100(count=1),
+    timeout= 1 * HOURS,
     )
-    print("🍡 model downloaded")
-
-    print("🍡 downloading text encoder")
-
-    T5Tokenizer.from_pretrained("google/t5-v1_1-xxl", legacy=False)
-    T5EncoderModel.from_pretrained("google/t5-v1_1-xxl")
-
-    model.commit()
-    print("🍡 text encoder downloaded")
+class Mochi:
+    # We can stack the `modal.build` and `modal.enter` decorators
+    # because we use the diffusers library, which caches the model to disk on its first run.
+    # This builds our model weights into the container image, ready for future use.
+    @modal.build()
+    @modal.enter()
+    def load_model(self):
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.pipe = MochiPipeline.from_pretrained(
+            "genmo/mochi-1-preview", 
+            torch_dtype=torch.bfloat16, 
+        )
+        self.pipe.enable_model_cpu_offload()
+        self.pipe.enable_vae_tiling()
 
+    @modal.method()
+    def generate(self, 
+        prompt, 
+        negative_prompt = "",
+        num_inference_steps = 200,
+        guidance_scale = 4.5,
+        num_frames = 19,
+    ):
+        frames = self.pipe(
+            prompt=prompt, 
+            negative_prompt=negative_prompt,
+            num_inference_steps=num_inference_steps, 
+            guidance_scale=guidance_scale,
+            num_frames=num_frames,
+        ).frames[0]
+        
+        # save to disk using prompt as filename
+        for char in string.punctuation:
+            prompt = prompt.replace(char, "")
+        prompt = prompt.replace(" ", "_")
+        prompt = prompt[:230] # since filenames can't be longer than 255 characters
+        mp4_name = str(int(time.time())) + "_" + prompt + ".mp4"
+        
+        export_to_video(frames, Path(OUTPUTS_PATH) /  mp4_name)
+        outputs.commit()
```

Comment:
This shouldn't be necessary given that you're returning immediately afterwards

Reply:
Before I added `.commit()`, I'd gotten a File Not Exists error from the local entrypoint when immediately downloading from the volume after function return. If a volume isn't fully committed on function return (without manually calling .commit), is that a modal bug?

Reply:
Oh, actually that's right -- we don't guarantee that. Always trips me up because I expect it to work that way.